rosbag2_cpp/src/rosbag2_cpp/writers/cache/message_cache.cpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <utility>
+
+#include "rosbag2_cpp/logging.hpp"
+#include "rosbag2_cpp/writers/cache/message_cache.hpp"
+
+namespace rosbag2_cpp
+{
+namespace writers
+{
+namespace cache
+{
+
+MessageCache::MessageCache(const uint64_t & max_buffer_size)
+{
+  primary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+  secondary_buffer_ = std::make_shared<MessageCacheBuffer>(max_buffer_size);
+}
+
+bool MessageCache::push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg)
+{
+  // While pushing, we keep track of inserted and dropped messages as well
+  bool pushed = false;
+  {
+    std::lock_guard<std::mutex> buffer_lock(buffer_mutex_);
+    pushed = primary_buffer_->push(msg);
+  }
+
+  allow_swap();
+
+  if (!pushed) {
+    elements_dropped_++;
```

Comment:
we should log at least on which topic (should be part of `msg`) is getting lost.